### PR TITLE
Fixes issue #103

### PR DESCRIPTION
**Pull request by sebglane**
This PR yields an executable `GuermondNeumannBC` which prints the following output:
```
                               Velocity convergence table
==============================================================================================
level    dt    cells dofs   hmax           L2                 H1               Linfty       
    5 5.00e-02  1024 8450 4.42e-02 1.174374e-02     - 7.747975e-02     - 2.008523e-02     - 
    5 2.50e-02  1024 8450 4.42e-02 8.459234e-04 -3.80 4.012410e-03 -4.27 1.439663e-03 -3.80 
    5 1.25e-02  1024 8450 4.42e-02 2.594778e-04 -1.70 1.629647e-03 -1.30 6.218237e-04 -1.21 
    5 6.25e-03  1024 8450 4.42e-02 7.592145e-05 -1.77 8.612342e-04 -0.92 3.423132e-04 -0.86 
    5 3.12e-03  1024 8450 4.42e-02 8.040314e-05  0.08 6.861911e-04 -0.33 2.388432e-04 -0.52 

                               Pressure convergence table
==============================================================================================
level    dt    cells dofs   hmax           L2                 H1               Linfty       
    5 5.00e-02  1024 1089 4.42e-02 1.146902e-01     - 3.703239e+00     - 8.414710e-01     - 
    5 2.50e-02  1024 1089 4.42e-02 3.467230e-03 -5.05 5.684240e-02 -6.03 1.036671e-02 -6.34 
    5 1.25e-02  1024 1089 4.42e-02 2.078839e-03 -0.74 3.547187e-02 -0.68 5.246254e-03 -0.98 
    5 6.25e-03  1024 1089 4.42e-02 1.707197e-03 -0.28 2.870705e-02 -0.31 3.841652e-03 -0.45 
    5 3.12e-03  1024 1089 4.42e-02 1.559528e-03 -0.13 2.609393e-02 -0.14 3.542803e-03 -0.12 
```

### CODE REVIEW

File: applications/GuermondNeumannBC.cc
Comment (j507) [inline applications/GuermondNeumannBC.cc:206]: This is the problem. By enforcing this boundary conditions on the pressure you are formulating another problem in the `phi` space due to how I constrain `phi` inside `NavierStokesProjection` according to the pressure and velocity spaces. The original Neumann problem translate into
```
+---------+------------+------------------------------------------+
| Boundary conditions of the Velocity entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 0       | Neumann    | ZeroTensorFunction<1, 2, double>         |
| 1       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 2       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 3       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Pressure entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| Unconstrained boundary ids: 0, 1, 2, 3                          |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Phi entity                           |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 1       | Neumann    | ZeroFunction<2, double>                  |
| 2       | Neumann    | ZeroFunction<2, double>                  |
| 3       | Neumann    | ZeroFunction<2, double>                  |
| 0       | Dirichlet  | ZeroFunction<2, double>                  |
+---------+------------+------------------------------------------+
```
but with your changes the problem becomes
```
+---------+------------+------------------------------------------+
| Boundary conditions of the Velocity entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 0       | Neumann    | ZeroTensorFunction<1, 2, double>         |
| 1       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 2       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 3       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Pressure entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 1       | Dirichlet  | GuermondNeumannBC::PressureExactSolution<2> |
| 2       | Dirichlet  | GuermondNeumannBC::PressureExactSolution<2> |
| 3       | Dirichlet  | GuermondNeumannBC::PressureExactSolution<2> |
| Unconstrained boundary ids: 0                                   |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Phi entity                           |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 0       | Dirichlet  | ZeroFunction<2, double>                  |
| 1       | Dirichlet  | ZeroFunction<2, double>                  |
| 2       | Dirichlet  | ZeroFunction<2, double>                  |
| 3       | Dirichlet  | ZeroFunction<2, double>                  |
+---------+------------+------------------------------------------+
```

**Review by j507 — CHANGES_REQUESTED** [review]

File: applications/GuermondNeumannBC.cc
Comment (sebglane) [reply applications/GuermondNeumannBC.cc:206]: Okay I understand. With the original code I get 
```
+---------+------------+------------------------------------------+
| Boundary conditions of the Velocity entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| 0       | Neumann    | ZeroTensorFunction<1, 2, double>         |
| 1       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 2       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
| 3       | Dirichlet  | GuermondNeumannBC::VelocityExactSolution<2> |
+---------+------------+------------------------------------------+
+---------+------------+------------------------------------------+
| Boundary conditions of the Pressure entity                      |
+---------+------------+------------------------------------------+
| Bdy. id |    Type    |              Function                    |
| Unconstrained boundary ids: 0, 1, 2, 3                          |
+---------+------------+------------------------------------------+
```

but the problem is that I get the runtime error

```
----------------------------------------------------
Exception on processing: 

--------------------------------------------------------
An error occurred in line <624> of file </home/sg/Documents/Projects/Git/RotatingMHD/source/entities_structs.cc> in function
    void RMHD::Entities::ScalarEntity<dim>::apply_boundary_conditions() [with int dim = 2]
The violated condition was: 
    boundary_conditions.regularity_guaranteed()
Additional information: 
    No boundary conditions were set for the "Pressure" entity
--------------------------------------------------------

Aborting!
----------------------------------------------------
```
because the pressure does not have any boundary condition. Should I apply a mean value constraint or should we resolve this?

Comment (j507) [reply applications/GuermondNeumannBC.cc:206]: > because the pressure does not have any boundary condition. Should I apply a mean value constraint or should we resolve this?

Applying a mean value constraint would formulate a different problem. We have to build a bypass of the regularity check for the pressure. As it's regularity can be ensured indirectly. A not so elegant solution is just to pass a flag to the constructor to indicate the bypass. A general scalar field, i.e. the temperature field, needs this regularity check so deleting is not really a solution.

**Comment by sebglane** [comment]
With the modification in 2943227, `GuermondNeumannBC.cc` is executable again and the results seem reasonable. Ready to go if you think so.

**Comment by j507** [comment]
I tried to replicate the [old results](https://github.com/j507/RotatingMHD/pull/47#issuecomment-754920007) but I get something else
```
                               Velocity convergence table
==============================================================================================
level    dt    cells  dofs    hmax           L2                 H1               Linfty       
    7 5.00e-02 16384 132098 1.10e-02 1.320967e-03     - 5.937734e-03     - 2.322636e-03     - 
    7 2.50e-02 16384 132098 1.10e-02 3.944570e-04 -1.74 2.202181e-03 -1.43 9.523212e-04 -1.29 
    7 1.25e-02 16384 132098 1.10e-02 1.242409e-04 -1.67 8.617592e-04 -1.35 3.641436e-04 -1.39 
    7 6.25e-03 16384 132098 1.10e-02 3.997212e-05 -1.64 3.561557e-04 -1.27 1.363815e-04 -1.42 
    7 3.12e-03 16384 132098 1.10e-02 2.564857e-05 -0.64 2.402366e-04 -0.57 1.067046e-04 -0.35 
    7 1.56e-03 16384 132098 1.10e-02 1.742542e-05 -0.56 1.482638e-04 -0.70 5.999823e-05 -0.83 
    7 7.81e-04 16384 132098 1.10e-02 2.688799e-05  0.63 1.882444e-04  0.34 7.414017e-05  0.31 

                               Pressure convergence table
==============================================================================================
level    dt    cells dofs    hmax           L2                 H1               Linfty       
    7 5.00e-02 16384 16641 1.10e-02 3.707286e-03     - 5.331211e-02     - 5.431266e-02     - 
    7 2.50e-02 16384 16641 1.10e-02 1.296654e-03 -1.52 2.608775e-02 -1.03 1.524138e-02 -1.83 
    7 1.25e-02 16384 16641 1.10e-02 4.074285e-04 -1.67 1.247370e-02 -1.06 4.481017e-03 -1.77 
    7 6.25e-03 16384 16641 1.10e-02 2.331934e-04 -0.81 6.922621e-03 -0.85 2.294989e-03 -0.97 
    7 3.12e-03 16384 16641 1.10e-02 4.064540e-04  0.80 5.802429e-03 -0.25 1.199593e-03 -0.94 
    7 1.56e-03 16384 16641 1.10e-02 2.760444e-04 -0.56 4.072635e-03 -0.51 6.270250e-04 -0.94 
    7 7.81e-04 16384 16641 1.10e-02 5.267589e-04  0.93 5.177419e-03  0.35 1.031357e-03  0.72 
```
Currently running it on master. I will keep you posted.

**Comment by sebglane** [comment]
> I tried to replicate the [old results](https://github.com/j507/RotatingMHD/pull/47#issuecomment-754920007) but I get something else
> 
> ```
>                                Velocity convergence table
> ==============================================================================================
> level    dt    cells  dofs    hmax           L2                 H1               Linfty       
>     7 5.00e-02 16384 132098 1.10e-02 1.320967e-03     - 5.937734e-03     - 2.322636e-03     - 
>     7 2.50e-02 16384 132098 1.10e-02 3.944570e-04 -1.74 2.202181e-03 -1.43 9.523212e-04 -1.29 
>     7 1.25e-02 16384 132098 1.10e-02 1.242409e-04 -1.67 8.617592e-04 -1.35 3.641436e-04 -1.39 
>     7 6.25e-03 16384 132098 1.10e-02 3.997212e-05 -1.64 3.561557e-04 -1.27 1.363815e-04 -1.42 
>     7 3.12e-03 16384 132098 1.10e-02 2.564857e-05 -0.64 2.402366e-04 -0.57 1.067046e-04 -0.35 
>     7 1.56e-03 16384 132098 1.10e-02 1.742542e-05 -0.56 1.482638e-04 -0.70 5.999823e-05 -0.83 
>     7 7.81e-04 16384 132098 1.10e-02 2.688799e-05  0.63 1.882444e-04  0.34 7.414017e-05  0.31 
> 
>                                Pressure convergence table
> ==============================================================================================
> level    dt    cells dofs    hmax           L2                 H1               Linfty       
>     7 5.00e-02 16384 16641 1.10e-02 3.707286e-03     - 5.331211e-02     - 5.431266e-02     - 
>     7 2.50e-02 16384 16641 1.10e-02 1.296654e-03 -1.52 2.608775e-02 -1.03 1.524138e-02 -1.83 
>     7 1.25e-02 16384 16641 1.10e-02 4.074285e-04 -1.67 1.247370e-02 -1.06 4.481017e-03 -1.77 
>     7 6.25e-03 16384 16641 1.10e-02 2.331934e-04 -0.81 6.922621e-03 -0.85 2.294989e-03 -0.97 
>     7 3.12e-03 16384 16641 1.10e-02 4.064540e-04  0.80 5.802429e-03 -0.25 1.199593e-03 -0.94 
>     7 1.56e-03 16384 16641 1.10e-02 2.760444e-04 -0.56 4.072635e-03 -0.51 6.270250e-04 -0.94 
>     7 7.81e-04 16384 16641 1.10e-02 5.267589e-04  0.93 5.177419e-03  0.35 1.031357e-03  0.72 
> ```
> 
> Currently running it on master. I will keep you posted.

Okay what do you think is the reason for this? I can only imagine that the errors might be computed differently if `VectorTools::integrate_difference` is used instead of interpolating the exact solution on the FE function space and computing the error by comparing vectors.

I will try to replicate the convergence table on my machine, but it would be helpful if you send me the input file which generate the output above.

**Comment by j507** [comment]
The `master` branch produces the same results as `fix_103`. Whatever is causing this happened in another merge. I will have to search for the culprit. Meanwhile feel free to merge this.

**Comment by sebglane** [comment]
> The `master` branch produces the same results as `fix_103`. Whatever is causing this happened in another merge. I will have to search for the culprit. Meanwhile feel free to merge this.

Let us not merge this unless the problem is fixed. The changes were minimal so far and are not necessarily required to be merged for my other PRs. I did another convergence test on a really fine mesh with a maximum CFL number of 2.

```
                               Velocity convergence table
==============================================================================================
level    dt    cells  dofs    hmax           L2                 H1               Linfty       
    8 1.00e-02 65536 526338 5.52e-03 1.218628e-04     - 1.143775e-03     - 5.725481e-04     - 
    8 5.00e-03 65536 526338 5.52e-03 1.230077e-04  0.01 9.196279e-04 -0.31 4.238432e-04 -0.43 
    8 2.50e-03 65536 526338 5.52e-03 8.665831e-05 -0.51 6.270741e-04 -0.55 2.626781e-04 -0.69 
    8 1.25e-03 65536 526338 5.52e-03 6.629196e-05 -0.39 4.633193e-04 -0.44 1.733251e-04 -0.60 
    8 6.25e-04 65536 526338 5.52e-03 6.274130e-05 -0.08 4.255919e-04 -0.12 1.513539e-04 -0.20 

                               Pressure convergence table
==============================================================================================
level    dt    cells dofs    hmax           L2                 H1               Linfty       
    8 1.00e-02 65536 66049 5.52e-03 2.793942e-03     - 2.230052e-02     - 7.849388e-03     - 
    8 5.00e-03 65536 66049 5.52e-03 1.825054e-03 -0.61 1.541061e-02 -0.53 3.643626e-03 -1.11 
    8 2.50e-03 65536 66049 5.52e-03 1.344578e-03 -0.44 1.414401e-02 -0.12 3.165391e-03 -0.20 
    8 1.25e-03 65536 66049 5.52e-03 1.058741e-03 -0.34 1.125499e-02 -0.33 2.428034e-03 -0.38 
    8 6.25e-04 65536 66049 5.52e-03 1.089469e-03  0.04 1.160903e-02  0.04 2.435659e-03  0.00 
```

**Comment by j507** [comment]
With the corrected tolerance values of #116 one obtains 
```
                               Velocity convergence table
==============================================================================================
level    dt    cells  dofs    hmax           L2                 H1               Linfty       
    7 5.00e-02 16384 132098 1.10e-02 1.332230e-03     - 5.934953e-03     - 2.431128e-03     - 
    7 2.50e-02 16384 132098 1.10e-02 4.364486e-04 -1.61 2.164054e-03 -1.46 9.042184e-04 -1.43 
    7 1.25e-02 16384 132098 1.10e-02 1.387462e-04 -1.65 8.334347e-04 -1.38 3.225033e-04 -1.49 
    7 6.25e-03 16384 132098 1.10e-02 4.338351e-05 -1.68 3.356971e-04 -1.31 1.151210e-04 -1.49 
    7 3.12e-03 16384 132098 1.10e-02 1.351288e-05 -1.68 1.388374e-04 -1.27 4.458773e-05 -1.37 
    7 1.56e-03 16384 132098 1.10e-02 4.231193e-06 -1.68 5.817881e-05 -1.25 1.711882e-05 -1.38 
    7 7.81e-04 16384 132098 1.10e-02 1.339325e-06 -1.66 2.462103e-05 -1.24 6.510449e-06 -1.39 

                               Pressure convergence table
==============================================================================================
level    dt    cells dofs    hmax           L2                 H1               Linfty       
    7 5.00e-02 16384 16641 1.10e-02 3.041071e-03     - 5.218148e-02     - 5.199298e-02     - 
    7 2.50e-02 16384 16641 1.10e-02 7.813870e-04 -1.96 2.351617e-02 -1.15 1.442508e-02 -1.85 
    7 1.25e-02 16384 16641 1.10e-02 2.563195e-04 -1.61 1.166905e-02 -1.01 4.239410e-03 -1.77 
    7 6.25e-03 16384 16641 1.10e-02 1.007239e-04 -1.35 6.194251e-03 -0.91 2.301475e-03 -0.88 
    7 3.12e-03 16384 16641 1.10e-02 4.101860e-05 -1.30 3.771291e-03 -0.72 1.335871e-03 -0.78 
    7 1.56e-03 16384 16641 1.10e-02 1.655232e-05 -1.31 2.905428e-03 -0.38 7.207308e-04 -0.89 
    7 7.81e-04 16384 16641 1.10e-02 6.876598e-06 -1.27 2.670068e-03 -0.12 3.624371e-04 -0.99 
```
which are the original results of #47.

**Review by j507 — APPROVED** [review]

**Pull request merged**
